Pregunta 1 :
Explica como el algoritmo K-Nearest Neighbors utiliza metricas de similitud para generar recomendaciones.
Indica dos ventajas y dos limitaciones del uso de KNN en un sistema de recomendacion con datos de peliculas

Pregunta 2 :
Selecciona dos usuarios al azar que hayan valorado al menos 5 peliculas en comun.
Calcula la similitud entre estos dos usuarios usando la correlacion de Pearson. Luego, interpreta el resultado
y discute que indicaria esta similitud para un sistema que busca recomendar peliculas.

Pregunta 3 :
Escoge una pelicula que tenga al menos 50 valoraciones y calcula su similitud con otras dos peliculas usando la similitud
coseno.Explica como podrias usar estos resultados para sugerir recomendaciones basadas en items.

Pregunta 4:
Implementa un sistema de recomendacion basado en KNN utilizando la distancia Euclideanda como metrica de similitud.
Divide el datased en conjuntos de entrenamineto y prueba.Luego , calcula el hit rate del modelo en el conjunto de prueba, explicando como el hit rate
refleja la efectividad del recomendador en este caso.

## Pregunta 1 : Explica como el algoritmo K-Nearest Neighbors utiliza metricas de similitud para generar recomendaciones. Indica dos ventajas y dos limitaciones del uso de KNN en un sistema de recomendacion con datos de peliculas.

###  Respuesta

KNN en sistemas de recomendación identifica elementos similares a un usuario o un ítem (película) en función de métricas de similitud, como la distancia euclidiana, la similitud de coseno, o la correlación de Pearson. En un sistema de recomendación de películas, el algoritmo puede buscar usuarios con perfiles similares o películas con características o ratings parecidos. Basándose en estas similitudes, KNN recomienda películas populares entre los usuarios más similares o películas con un perfil similar a las que un usuario ya ha disfrutado.

** Ventajas de KNN:

- Fácil de interpretar: Es intuitivo y fácil de implementar, pues recomienda películas en función de la cercanía a los ítems o usuarios más similares.
- No requiere entrenamiento pesado: Al ser un modelo basado en instancias, no necesita pre-entrenamiento intensivo.

** Limitaciones de KNN:

- Escalabilidad: Puede volverse computacionalmente costoso en datasets grandes, ya que debe calcular la distancia entre muchos pares de usuarios o películas.
- Escasez de datos: La precisión puede ser baja si hay poca información (ratings) compartida entre usuarios, lo que afecta la calidad de las recomendaciones.

## Configuracion del entorno

In [1]:
# Descargar el dataset MovieLens 1M
!curl -o dataset.zip "https://files.grouplens.org/datasets/movielens/ml-1m.zip"
!unzip dataset.zip
!ls -la

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 5778k  100 5778k    0     0  6789k      0 --:--:-- --:--:-- --:--:-- 6782k
Archive:  dataset.zip
   creating: ml-1m/
  inflating: ml-1m/movies.dat        
  inflating: ml-1m/ratings.dat       
  inflating: ml-1m/README            
  inflating: ml-1m/users.dat         
total 5800
drwxr-xr-x 1 root root    4096 Oct 29 00:47 .
drwxr-xr-x 1 root root    4096 Oct 29 00:42 ..
drwxr-xr-x 4 root root    4096 Oct 25 13:20 .config
-rw-r--r-- 1 root root 5917549 Oct 29 00:47 dataset.zip
drwxr-x--- 2 root root    4096 Jan 29  2016 ml-1m
drwxr-xr-x 1 root root    4096 Oct 25 13:20 sample_data


In [2]:

import pandas as pd
import matplotlib.pyplot as plt


In [3]:
# Cargar los datasets de calificaciones y películas
ratings = pd.read_csv('ml-1m/ratings.dat', sep='::', header=None, engine='python',
                      names=['userId', 'movieId', 'rating', 'timestamp'], encoding='latin-1')
movies  = pd.read_csv('ml-1m/movies.dat', sep='::',  header=None, engine='python',
                      names=['movieId', 'title', 'genres'], encoding='latin-1')

In [4]:

# Unir los datasets en base a 'movieId' para agregar los títulos
user_item_rating = pd.merge(ratings, movies[['movieId', 'title']], on='movieId')

# Seleccionar solo las columnas necesarias
user_item_rating = user_item_rating[['userId', 'title', 'rating']]

# Ordenar los datos por 'userId'
user_item_rating.sort_values(by='userId', inplace=True)

# Mostrar las primeras filas para verificar
user_item_rating.head()


<ipython-input-4-79d1f1918971>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  user_item_rating.sort_values(by='userId', inplace=True)


userId                                   title  rating
0        1  One Flew Over the Cuckoo's Nest (1975)       5
29       1                   Close Shave, A (1995)       3
30       1                             Antz (1998)       4
31       1                Girl, Interrupted (1999)       4
32       1                         Hercules (1997)       4

Pregunta 2 : Selecciona dos usuarios al azar que hayan valorado al menos 5 peliculas en comun. Calcula la similitud entre estos dos usuarios usando la correlacion de Pearson. Luego, interpreta el resultado y discute que indicaria esta similitud para un sistema que busca recomendar peliculas.

In [6]:
from itertools import combinations
df = pd.DataFrame(user_item_rating)

# Paso 1: Crear una tabla pivote con los usuarios y las películas que calificaron
pivot = df.pivot_table(index='userId', columns='title', values='rating', aggfunc='first')

# Paso 2: Encontrar pares de usuarios que tengan al menos 5 películas en común
user_pairs = []
for user1, user2 in combinations(pivot.index, 2):
    # Contar cuántas películas ambos usuarios han calificado
    common_movies = pivot.loc[user1].notna() & pivot.loc[user2].notna()
    if common_movies.sum() >= 5:
        user_pairs.append((user1, user2))

KeyboardInterrupt: 

El codigo de arriba se pauso xq demoraba pero se puede encontrar todas las combinaciones y validar

In [11]:
user_pairs

[(1, 2),
 (1, 3),
 (1, 5),
 (1, 6),
 (1, 8),
 (1, 9),
 (1, 10),
 (1, 11),
 (1, 13),
 (1, 15),
 (1, 17),
 (1, 18),
 (1, 19),
 (1, 21),
 (1, 22),
 (1, 23),
 (1, 24),
 (1, 25),
 (1, 26),
 (1, 28),
 (1, 30),
 (1, 33),
 (1, 34),
 (1, 35),
 (1, 36),
 (1, 38),
 (1, 39),
 (1, 42),
 (1, 44),
 (1, 45),
 (1, 48),
 (1, 49),
 (1, 51),
 (1, 53),
 (1, 55),
 (1, 56),
 (1, 58),
 (1, 59),
 (1, 62),
 (1, 65),
 (1, 67),
 (1, 68),
 (1, 70),
 (1, 72),
 (1, 73),
 (1, 75),
 (1, 76),
 (1, 78),
 (1, 80),
 (1, 82),
 (1, 83),
 (1, 86),
 (1, 88),
 (1, 89),
 (1, 90),
 (1, 92),
 (1, 93),
 (1, 95),
 (1, 96),
 (1, 97),
 (1, 99),
 (1, 100),
 (1, 102),
 (1, 105),
 (1, 107),
 (1, 109),
 (1, 111),
 (1, 112),
 (1, 114),
 (1, 117),
 (1, 118),
 (1, 119),
 (1, 120),
 (1, 121),
 (1, 123),
 (1, 127),
 (1, 129),
 (1, 131),
 (1, 132),
 (1, 133),
 (1, 134),
 (1, 135),
 (1, 136),
 (1, 137),
 (1, 139),
 (1, 142),
 (1, 143),
 (1, 146),
 (1, 147),
 (1, 148),
 (1, 149),
 (1, 150),
 (1, 151),
 (1, 152),
 (1, 156),
 (1, 157),
 (1, 159),


## Optimizado si solo quiero un par

In [59]:
pivot = df.pivot_table(index='userId', columns='title', values='rating', aggfunc='first')

# Paso 2: Encontrar el primer par de usuarios que tengan al menos 5 películas en común
first_pair = None
for user1, user2 in combinations(pivot.index, 2):
    # Contar cuántas películas ambos usuarios han calificado
    common_movies = pivot.loc[user1].notna() & pivot.loc[user2].notna()
    if common_movies.sum() >= 5:
        first_pair = (user1, user2)
        break  # Detener el bucle al encontrar el primer par

# Imprimir el primer par de usuarios que calificaron al menos 5 películas en común
print("Primer par de usuarios con al menos 5 películas en común:", first_pair)

Primer par de usuarios con al menos 5 películas en común: (1, 2)


In [60]:
first_pair

(1, 2)

In [17]:
# Estos usuarios salio seleccionada que tienen 5 titulos en comun
Titulos_1=list(user_item_rating[user_item_rating['userId']==1].title)
Titulos_2=list(user_item_rating[user_item_rating['userId']==1332].title)

## Validamos que tengan 5 titulos en comun
comun = set(Titulos_1) & set(Titulos_2)
print("Tienen en comun: ",len(comun), " titulos que valoraron")

Tienen en comun:  6  titulos que valoraron


In [29]:
Usuario1=user_item_rating[(user_item_rating['userId']==1) & (user_item_rating['title'].isin(comun)==True)].sort_values('title')
Usuario1

userId                                   title  rating
15       1                        Airplane! (1980)       4
7        1               Christmas Story, A (1983)       5
0        1  One Flew Over the Cuckoo's Nest (1975)       5
5        1              Princess Bride, The (1987)       3
40       1                        Toy Story (1995)       5
50       1                      Toy Story 2 (1999)       4

In [28]:
Usuario2=user_item_rating[(user_item_rating['userId']==1332) & (user_item_rating['title'].isin(comun)==True)].sort_values('title')
Usuario2

userId                                   title  rating
219609    1332                        Airplane! (1980)       5
219589    1332               Christmas Story, A (1983)       5
219595    1332  One Flew Over the Cuckoo's Nest (1975)       5
219596    1332              Princess Bride, The (1987)       4
219581    1332                        Toy Story (1995)       5
219598    1332                      Toy Story 2 (1999)       5

In [30]:
ratings_user1 = Usuario1['rating'].values
print(ratings_user1)
ratings_user2 = Usuario2['rating'].values
print(ratings_user2)

[4 5 5 3 5 4]
[5 5 5 4 5 5]


In [32]:
from scipy.stats import pearsonr
if len(ratings_user1) > 1:  # Pearson requiere al menos 2 puntos para calcular la correlación
    correlation, _ = pearsonr(ratings_user1, ratings_user2)
    print("Similitud de Pearson entre el usuario 1 y el usuario 1332:", correlation)
else:
    print("No hay suficientes calificaciones en común para calcular la correlación de Pearson.")

Similitud de Pearson entre el usuario 1 y el usuario 1332: 0.7999999999999998


La similitud de Pearson calculada entre el usuario 1 y el usuario 1332 es aproximadamente 0.80. Este resultado indica una fuerte correlación positiva en las calificaciones de las películas que ambos usuarios han evaluado. Dado que han calificado al menos 5 películas en común, esta medida sugiere que:

- Ambos usuarios tienden a dar calificaciones altas o bajas a las mismas películas de manera consistente. Por ejemplo, si el usuario 1 califica una película como 5 (muy buena), es muy probable que el usuario 1332 también le dé una calificación alta a esa misma película. Esta tendencia sugiere que ambos comparten gustos similares

- Dado que hay una fuerte correlación entre las calificaciones de los dos usuarios, el sistema de recomendación puede utilizar esta información para sugerir películas que el usuario 1 ha calificado positivamente, pero que el usuario 1332 aún no ha visto. Esto es especialmente útil en plataformas de streaming donde los usuarios buscan contenido relevante.
- Por ejemplo, si el usuario 1 ha calificado una nueva película con un puntaje alto, el sistema podría recomendar esa película al usuario 1332, confiando en que probablemente le gustará.
- Esta similitud es un ejemplo del enfoque de filtrado colaborativo, que se basa en las preferencias de los usuarios similares para hacer recomendaciones. El sistema puede construir un perfil de usuario para el usuario 1332, utilizando no solo sus propias calificaciones, sino también las calificaciones de otros usuarios con patrones similares, como el usuario 1.

Pregunta 3 : Escoge una pelicula que tenga al menos 50 valoraciones y calcula su similitud con otras dos peliculas usando la similitud coseno.Explica como podrias usar estos resultados para sugerir recomendaciones basadas en items.

In [39]:
cantidad_valoraciones=user_item_rating.groupby('title').count()
cantidad_valoraciones = cantidad_valoraciones.sort_values('rating', ascending=False)
print(cantidad_valoraciones.head())

                                                    userId  rating
title                                                             
American Beauty (1999)                                3428    3428
Star Wars: Episode IV - A New Hope (1977)             2991    2991
Star Wars: Episode V - The Empire Strikes Back ...    2990    2990
Star Wars: Episode VI - Return of the Jedi (1983)     2883    2883
Jurassic Park (1993)                                  2672    2672


In [42]:
peliculas_mas_50_valoraciones = cantidad_valoraciones[cantidad_valoraciones['rating'] >= 50]
peliculas_mas_50_valoraciones

userId  rating
title                                                             
American Beauty (1999)                                3428    3428
Star Wars: Episode IV - A New Hope (1977)             2991    2991
Star Wars: Episode V - The Empire Strikes Back ...    2990    2990
Star Wars: Episode VI - Return of the Jedi (1983)     2883    2883
Jurassic Park (1993)                                  2672    2672
...                                                    ...     ...
Blow-Out (La Grande Bouffe) (1973)                      50      50
Washington Square (1997)                                50      50
Associate, The (1996)                                   50      50
Audrey Rose (1977)                                      50      50
Heidi Fleiss: Hollywood Madam (1995)                    50      50

[2514 rows x 2 columns]

In [44]:
##Pelicula escogida Jurassic Park (1993)
print("Película seleccionada:", peliculas_mas_50_valoraciones.index[4])
pelicula_seleccionada=peliculas_mas_50_valoraciones.index[4]

Película seleccionada: Jurassic Park (1993)


In [48]:
from sklearn.metrics.pairwise import cosine_similarity

# Crear una tabla de usuario-película
user_movie_matrix = user_item_rating.pivot_table(index='userId', columns='title', values='rating')
user_movie_matrix

title   $1,000,000 Duck (1971)  'Night Mother (1986)  \
userId                                                 
1                          NaN                   NaN   
2                          NaN                   NaN   
3                          NaN                   NaN   
4                          NaN                   NaN   
5                          NaN                   NaN   
...                        ...                   ...   
6036                       NaN                   3.0   
6037                       NaN                   NaN   
6038                       NaN                   NaN   
6039                       NaN                   NaN   
6040                       NaN                   NaN   

title   'Til There Was You (1997)  'burbs, The (1989)  \
userId                                                  
1                             NaN                 NaN   
2                             NaN                 NaN   
3                             NaN                 NaN   
4                             NaN                 NaN   
5                             NaN                 NaN   
...                           ...                 ...   
6036                          NaN                 NaN   
6037                          NaN                 NaN   
6038                          NaN                 NaN   
6039                          NaN                 NaN   
6040                          NaN                 NaN   

title   ...And Justice for All (1979)  1-900 (1994)  \
userId                                                
1                                 NaN           NaN   
2                                 NaN           NaN   
3                                 NaN           NaN   
4                                 NaN           NaN   
5                                 NaN           NaN   
...                               ...           ...   
6036                              NaN           NaN   
6037                              NaN           NaN   
6038                              NaN           NaN   
6039                              NaN           NaN   
6040                              NaN           NaN   

title   10 Things I Hate About You (1999)  101 Dalmatians (1961)  \
userId                                                             
1                                     NaN                    NaN   
2                                     NaN                    NaN   
3                                     NaN                    NaN   
4                                     NaN                    NaN   
5                                     NaN                    NaN   
...                                   ...                    ...   
6036                                  2.0                    4.0   
6037                                  NaN                    NaN   
6038                                  NaN                    NaN   
6039                                  NaN                    NaN   
6040                                  NaN                    NaN   

title   101 Dalmatians (1996)  12 Angry Men (1957)  ...  \
userId                                              ...   
1                         NaN                  NaN  ...   
2                         NaN                  NaN  ...   
3                         NaN                  NaN  ...   
4                         NaN                  NaN  ...   
5                         NaN                  NaN  ...   
...                       ...                  ...  ...   
6036                      NaN                  NaN  ...   
6037                      NaN                  4.0  ...   
6038                      NaN                  NaN  ...   
6039                      NaN                  NaN  ...   
6040                      NaN                  5.0  ...   

title   Young Poisoner's Handbook, The (1995)  Young Sherlock Holmes (1985)  \
userId                                                                        
1 

Considerando que se pueden seleccionar peliculas que tengan menos de 50 valoraciones , usaremos peliculas que esten dentro de la misma lista de la pelicula seleccionada esto con el fin de poder tener mas informacion sobre usarios y no escoger una pelicula con pocas valoraciones y ser comparada con una que tenga muchas valoraciones.

In [49]:
# Seleccionar dos películas para comparar con la película seleccionada (Se escogio el puesto 11 y 21)
pelicula_1 = peliculas_mas_50_valoraciones.index[10]
pelicula_2 = peliculas_mas_50_valoraciones.index[20]

# Extraer los vectores de las películas
vector_pelicula_seleccionada = user_movie_matrix[pelicula_seleccionada].fillna(0).values.reshape(1, -1)
vector_pelicula_1 = user_movie_matrix[pelicula_1].fillna(0).values.reshape(1, -1)
vector_pelicula_2 = user_movie_matrix[pelicula_2].fillna(0).values.reshape(1, -1)

# Calcular la similitud coseno
similitud_pelicula_1 = cosine_similarity(vector_pelicula_seleccionada, vector_pelicula_1)[0][0]
similitud_pelicula_2 = cosine_similarity(vector_pelicula_seleccionada, vector_pelicula_2)[0][0]

print(f"Similitud coseno entre {pelicula_seleccionada} y {pelicula_1}: {similitud_pelicula_1}")
print(f"Similitud coseno entre {pelicula_seleccionada} y {pelicula_2}: {similitud_pelicula_2}")

Similitud coseno entre Jurassic Park (1993) y Men in Black (1997): 0.7286190494818522
Similitud coseno entre Jurassic Park (1993) y E.T. the Extra-Terrestrial (1982): 0.5551648349661475


- Los resultados de similitud coseno indican que Jurassic Park y Men in Black son más similares (0.729) en comparación con E.T. the Extra-Terrestrial (0.555). Esto significa que los usuarios que han calificado Jurassic Park tienden a calificar Men in Black de manera similar, lo que sugiere que ambos pueden compartir características como género, trama o estilo cinematográfico.
Recomendaciones Directas:

- Si un usuario ha calificado Jurassic Park de manera positiva (por ejemplo, un 4 o un 5), el sistema de recomendación podría sugerirle Men in Black como una opción de visualización, dado que la similitud coseno es alta. La idea es que el usuario probablemente disfrute de la película recomendada debido a su relación con una película que ya le gusta.
Construcción de un Perfil de Usuario:

- Los sistemas de recomendación pueden construir perfiles de usuario basados en las películas que han calificado. Si un usuario tiene un historial de calificaciones similares a las de otros usuarios que también han visto Jurassic Park, el sistema puede utilizar esta información para recomendar películas como Men in Black, aprovechando la correlación de calificaciones entre las películas.-

Pregunta 4: Implementa un sistema de recomendacion basado en KNN utilizando la distancia Euclideanda como metrica de similitud. Divide el datased en conjuntos de entrenamineto y prueba.Luego , calcula el hit rate del modelo en el conjunto de prueba, explicando como el hit rate refleja la efectividad del recomendador en este caso.

In [51]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import NearestNeighbors
from sklearn.metrics import pairwise_distances
import numpy as np

pivot = user_item_rating.pivot_table(index='userId', columns='title', values='rating', fill_value=0)
pivot

title   $1,000,000 Duck (1971)  'Night Mother (1986)  \
userId                                                 
1                          0.0                   0.0   
2                          0.0                   0.0   
3                          0.0                   0.0   
4                          0.0                   0.0   
5                          0.0                   0.0   
...                        ...                   ...   
6036                       0.0                   3.0   
6037                       0.0                   0.0   
6038                       0.0                   0.0   
6039                       0.0                   0.0   
6040                       0.0                   0.0   

title   'Til There Was You (1997)  'burbs, The (1989)  \
userId                                                  
1                             0.0                 0.0   
2                             0.0                 0.0   
3                             0.0                 0.0   
4                             0.0                 0.0   
5                             0.0                 0.0   
...                           ...                 ...   
6036                          0.0                 0.0   
6037                          0.0                 0.0   
6038                          0.0                 0.0   
6039                          0.0                 0.0   
6040                          0.0                 0.0   

title   ...And Justice for All (1979)  1-900 (1994)  \
userId                                                
1                                 0.0           0.0   
2                                 0.0           0.0   
3                                 0.0           0.0   
4                                 0.0           0.0   
5                                 0.0           0.0   
...                               ...           ...   
6036                              0.0           0.0   
6037                              0.0           0.0   
6038                              0.0           0.0   
6039                              0.0           0.0   
6040                              0.0           0.0   

title   10 Things I Hate About You (1999)  101 Dalmatians (1961)  \
userId                                                             
1                                     0.0                    0.0   
2                                     0.0                    0.0   
3                                     0.0                    0.0   
4                                     0.0                    0.0   
5                                     0.0                    0.0   
...                                   ...                    ...   
6036                                  2.0                    4.0   
6037                                  0.0                    0.0   
6038                                  0.0                    0.0   
6039                                  0.0                    0.0   
6040                                  0.0                    0.0   

title   101 Dalmatians (1996)  12 Angry Men (1957)  ...  \
userId                                              ...   
1                         0.0                  0.0  ...   
2                         0.0                  0.0  ...   
3                         0.0                  0.0  ...   
4                         0.0                  0.0  ...   
5                         0.0                  0.0  ...   
...                       ...                  ...  ...   
6036                      0.0                  0.0  ...   
6037                      0.0                  4.0  ...   
6038                      0.0                  0.0  ...   
6039                      0.0                  0.0  ...   
6040                      0.0                  5.0  ...   

title   Young Poisoner's Handbook, The (1995)  Young Sherlock Holmes (1985)  \
userId                                                                        
1 

In [68]:
def Random_Holdout(df, test_size=0.25):
    """Divide los datos asegurando que cada usuario tenga al menos una interacción en el conjunto de entrenamiento"""
    # Seleccionar al azar un porcentaje de interacciones por usuario para entrenamiento
    train_df = df.groupby('userId').apply(lambda x: x.sample(frac=1 - test_size, random_state=42)).reset_index(drop=True)
    # El resto va al conjunto de prueba
    test_df = pd.concat([df, train_df]).drop_duplicates(keep=False)

    return train_df, test_df

# Aplicar la función al dataset filtrado
train_df, test_df = Random_Holdout(user_item_rating, test_size=0.3)

<ipython-input-68-2454e6bf9878>:4: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  train_df = df.groupby('userId').apply(lambda x: x.sample(frac=1 - test_size, random_state=42)).reset_index(drop=True)


In [69]:
train_df

userId                            title  rating
0            1       Saving Private Ryan (1998)       5
1            1             Pleasantville (1998)       3
2            1                       Big (1988)       4
3            1                 Toy Story (1995)       5
4            1                 Airplane! (1980)       4
...        ...                              ...     ...
700098    6040                 Spartacus (1960)       5
700099    6040                    Batman (1989)       2
700100    6040  Man with Two Brains, The (1983)       1
700101    6040               Grand Hotel (1932)       5
700102    6040                   Vertigo (1958)       4

[700103 rows x 3 columns]

In [70]:
test_df

userId                                title  rating
30            1                          Antz (1998)       4
35            1  Hunchback of Notre Dame, The (1996)       4
38            1              Sixth Sense, The (1999)       4
43            1     Run Lola Run (Lola rennt) (1998)       4
47            1         To Kill a Mockingbird (1962)       4
...         ...                                  ...     ...
999981     6040                  Out of Sight (1998)       3
999978     6040                       Network (1976)       4
999976     6040           Conan the Barbarian (1982)       1
999983     6040                     Quiz Show (1994)       2
1000208    6040    E.T. the Extra-Terrestrial (1982)       4

[300106 rows x 3 columns]

In [73]:
!pip install surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 154.4/154.4 kB 5.5 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.4-cp310-cp310-linux_x86_64.whl size=2357294 sha256=e54640b1a4346d1d90fa9b36775f16e49eac25558a02d5a00e2caa761e55fa4a
  Stored in directory: /root/.cache/pip/wheels/4b/3f/df/6acbf0a40397d9bf3ff97f582cc22fb9ce66adde75bc71fd54
Successfully built scikit-surprise


In [74]:
from surprise import Dataset, Reader, KNNBasic, accuracy

# Definir el rango de calificaciones
reader = Reader(rating_scale=(1, 5))

# Cargar datos de entrenamiento
trainset = Dataset.load_from_df(train_df[['userId', 'title', 'rating']], reader).build_full_trainset()

# Crear conjunto de prueba
testset = list(test_df[['userId', 'title', 'rating']].itertuples(index=False, name=None))


In [75]:

# Configurar opciones del modelo KNN
sim_options = {
    'name': 'msd',  # Mean Squared Difference (distancia euclidiana)
    'user_based': True  # Filtrado Usuario-Usuario
}

# Crear el modelo KNN
knn = KNNBasic(k=50, sim_options=sim_options)

# Entrenar el modelo
knn.fit(trainset)


Computing the msd similarity matrix...
Done computing similarity matrix.


In [76]:

# Hacer predicciones sobre el conjunto de prueba
predictions = knn.test(testset)

# Calcular el RMSE
rmse = accuracy.rmse(predictions)
print(f"RMSE del modelo: {rmse:.4f}")


RMSE: 0.9278
RMSE del modelo: 0.9278


In [78]:
predictions_df = pd.DataFrame(predictions, columns=['userId', 'movieId', 'real_rating', 'predicted_rating', 'details'])

## Hit Rate Global

In [79]:
# Definir umbral de relevancia
relevant_threshold = 4.0

# Inicializar contadores
hits = 0
total_relevant = 0

# Iterar sobre todas las predicciones
for _, row in predictions_df.iterrows():
    real_rating = row['real_rating']
    predicted_rating = row['predicted_rating']

    # Considerar películas con calificación real ≥ umbral como relevantes
    if real_rating >= relevant_threshold:
        total_relevant += 1
        # Si la predicción también es ≥ umbral, es un acierto
        if predicted_rating >= relevant_threshold:
            hits += 1

# Calcular el Hit Rate Global
hit_rate_global = hits / total_relevant if total_relevant > 0 else 0
print(f"Hit Rate Global: {hit_rate_global:.2f}")

Hit Rate Global: 0.46


## Hit Rate por usuario

In [80]:

# Inicializar lista para almacenar hit rates por usuario
hit_rates = []

# Agrupar predicciones por usuario
for user_id, group in predictions_df.groupby('userId'):
    hits = 0
    total_relevant = 0

    for _, row in group.iterrows():
        real_rating = row['real_rating']
        predicted_rating = row['predicted_rating']

        # Considerar películas relevantes
        if real_rating >= relevant_threshold:
            total_relevant += 1
            if predicted_rating >= relevant_threshold:
                hits += 1

    # Calcular Hit Rate para el usuario
    if total_relevant > 0:
        hit_rate_user = hits / total_relevant
        hit_rates.append(hit_rate_user)

# Calcular el Hit Rate promedio
hit_rate_avg_user = sum(hit_rates) / len(hit_rates) if len(hit_rates) > 0 else 0
print(f"Hit Rate promedio por usuario: {hit_rate_avg_user:.2f}")

Hit Rate promedio por usuario: 0.48


- Eficiencia del Modelo: Un Hit Rate global del 46% sugiere que el modelo tiene una capacidad razonable para presentar recomendaciones que los usuarios encuentran útiles. Esto implica que el sistema puede identificar patrones en las preferencias de los usuarios y, por ende, hacer recomendaciones que son relevantes para una parte significativa de ellos.

- Relevancia Promedio: El Hit Rate promedio por usuario de 48% sugiere que, a nivel individual, los usuarios tienden a interactuar positivamente con casi la mitad de las recomendaciones. Esto refuerza la idea de que el modelo tiene una buena comprensión de las preferencias del usuario, aunque hay espacio para mejorar.

- Áreas de Mejora: A pesar de los resultados positivos, un Hit Rate de 46% indica que hay un 54% de las recomendaciones que no son bien recibidas. Esto representa una oportunidad para ajustar el modelo, mejorar la calidad de las recomendaciones y optimizar el sistema para que sea más preciso. Se podrían explorar técnicas adicionales, como la incorporación de más datos o la personalización de las recomendaciones según los intereses específicos de cada usuario.